### PR TITLE
Add `pfx` HTTPS option

### LINKED ISSUE #1306: Better HTTP(S) Request options documentation
#### What would you like to discuss?

Since we made some modification to the `options`, this statement
```
options
Type: object

Any of the https.request options.
```
from the [documentation](https://github.com/sindresorhus/got#options) is no more valid.

Maybe the individual `https.Request` options should be documented on the got page.

<hr />

Node.JS options (checked ones are those already documented by Got)

From `http.request`
- [x] agent
- [x] auth
- [x] createConnection (not present in README.md)
- [ ] defaultPort
- [x] family
- [x] headers
- [ ] host
- [ ] hostname
- [ ] insecureHTTPParser
- [x] localAddress
- [x] lookup (not present in README.md)
- [ ] maxHeaderSize
- [x] method
- [x] path (might be reviewed)
- [x] port (not present in README.md)
- [x] protocol (not present in README.md)
- [ ] setHost
- [ ] socketPath
- [x] timeout

From `tls.createSecureContext` (documented `https.request`)
- [x] ca
- [x] cert
- [ ] ciphers (WIP #1435)
- [ ] clientCertEngine
- [ ] crl
- [ ] dhparam
- [ ] ecdhCurve
- [ ] honorCipherOrder (WIP #1435)
- [x] key
- [x] passphrase
- [x] pfx (#1364)
- [x] rejectUnauthorized
- [ ] secureOptions
- [ ] secureProtocol
- [ ] servername
- [ ] sessionIdContext
- [ ] highWaterMark
- [x] checkServerIdentity (unducumented by Node.JS)

From `tls.createSecureContext` (but undocumented in `https.request`)
- [ ] minVersion (WIP #1435)
- [ ] maxVersion (WIP #1435)

### PR DESCRIPTION
**Pull request by markdboyd**
Fixes #1306

Adds support for the `pfx` option from TLS `SecureContextOptions` to the `HTTPSOptions` accepted by got.  

I tried to follow the existing convention of `passphrase` where `options.pfx` and `options.https.pfx` are both allowed, but using `options.pfx` prints a deprecation warning. 

~I also added a basic test of the deprecation warning~. I removed this test because it was failing since I wasn't providing a real PFX fixture. It also wasn't a very meaningful test.

I wasn't sure what other tests to add but I'm happy to add any that are requested.

#### Checklist

- [x] I have read the documentation.
- [x] I have included a pull request description of my changes.
- [x] I have included some tests.
- [x] If it's a new feature, I have included documentation updates.

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -985,6 +985,18 @@ Type: `string`
 
 The passphrase to decrypt the `options.https.key` (if different keys have different passphrases refer to `options.https.key` documentation).
 
+##### https.pfx
+
+Type: `string | Buffer | (string | Buffer | object)[]`
+
+PFX or PKCS12 encoded private key and certificate chain. `pfx` is an alternative to providing `options.https.key` and `options.https.certificate` individually. PFX is usually encrypted, if it is, `options.https.passphrase` will be used to decrypt it.
```
Comment (markdboyd) [inline readme.md:992]: This documentation more or less repeats the summary of `pfx` from https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_options. Is that okay?

File: source/core/index.ts [hunk 2]
```diff
@@ -684,6 +685,7 @@ export default class Request extends Duplex implements RequestEvents<Request> {
 			assert.any([is.string, is.object, is.array, is.undefined], options.https.key);
 			assert.any([is.string, is.object, is.array, is.undefined], options.https.certificate);
 			assert.any([is.string, is.undefined], options.https.passphrase);
+			assert.any([is.string, is.array, is.undefined], options.https.pfx);
```
Comment (markdboyd) [inline source/core/index.ts:688]: Is there a way to assert that a Buffer is allowed? It doesn't seem captured by these other types?

File: readme.md [hunk 3]
```diff
@@ -985,6 +985,18 @@ Type: `string`
...
+
+Multiple PFX can be be provided as an array of unencrypted buffers or an array of objects like:
+
+`{buf: <string|buffer>[, passphrase: <string>]}`
```
Comment (sindresorhus) [inline readme.md:996]: TypeScript notation is clearer:

```suggestion
`
{
	buffer: string | Buffer,
	passphrase?: string
}
`
```

File: source/core/index.ts (hunk 2 above)
Comment (sindresorhus) [reply source/core/index.ts:688]: `is.buffer`? https://github.com/sindresorhus/is#buffervalue

File: readme.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (markdboyd) [reply readme.md]: 335641bb4cde6813e2beb874ede850c0f933866d

File: source/core/index.ts (hunk 2 above)
Comment (markdboyd) [reply source/core/index.ts]: b039a37707d1d2774282205c290997493234b5b1

**Comment by markdboyd** [comment]
@sindresorhus Thanks for the feedback! I think this is ready for re-review

File: readme.md [hunk 4]
```diff
@@ -985,7 +985,24 @@ Type: `string`
 
 The passphrase to decrypt the `options.https.key` (if different keys have different passphrases refer to `options.https.key` documentation).
 
-##### Examples for `https.key`, `https.certificate` and `https.passphrase`
+##### https.pfx
+
+Type: `string | Buffer | (string | Buffer | object)[]`
```
Comment (sindresorhus) [inline readme.md:990]:
```suggestion
Type: `string | Buffer | Array<string | Buffer | object>`
```

This is more readable.

Comment (sindresorhus) [reply readme.md:990]: Why is `object` only supported in an array? Is that correct or a typo?

Comment (markdboyd) [reply readme.md]: And yes, it is correct that `object` is only supported in the array form. I updated the README to clarify

**Comment by Giotino** [comment]
This PR is missing the "HTTPS options restore" after the request (that has been added recently).

In case of retry a `DeprecationWarning` is emitted, in order to prevent this you need to add `pfx` to this section https://github.com/sindresorhus/got/blob/master/source/core/index.ts#L2391

Related issue: #1389 
Related PR: #1391

**Comment by Giotino** [comment]
> Yeah, sorry. I had started to work on adding the requested test, but I was having trouble creating a working PFX fixture using the instructions that @szmarczak linked. I can take another crack at it though

@markdboyd I've set-up in the "HTTPS Tests" PR a test for the PFX option.
[`13fcbdc` (#1430)](https://github.com/sindresorhus/got/pull/1430/commits/13fcbdc6c75c1e2e19f64d9e3a27fbab8fde243e)

Currently it uses the "hidden pfx option" which is deprecated.

**Review by szmarczak — APPROVED** [review]

**Comment by markdboyd** [comment]
Thanks @Giotino  and @szmarczak !

**Pull request merged**